piv/key.go

```diff
@@ -198,19 +198,29 @@ type verifier struct {
 }
 
 func (v *verifier) Verify(attestationCert, slotCert *x509.Certificate) (*Attestation, error) {
-	root := v.Root
-	if root == nil {
-		ca, err := yubicoCA()
+	var roots []*x509.Certificate
```

Comment:
The x509 package actually already has builtin support for validating a certificate against multiple roots. I think it'd be a good simplification to see if the upstream methods work. Does using a `x509.CertPool` instead of a `x509.Certificate` correctly validate? If so can we update the code to be something like:

```
o := x509.VerifyOptions{KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageAny}}
o.Roots = v.Roots
if o.Roots != nil {
    cas, err := yubicoRoots()
    if err != nil {
        // ...
    }
    o.Roots = cas
}

o.Intermediates := x509.NewCertPool()
o.Intermediates.AddCert(attestationCert)

if err := attestationCert.Verify(o); err != nil {
    // ...
}
return parseAttestation(slotCert)
```

I'm also happy to send that myself.

Reply:
Yes, the `x509.Certificate.Verify()` function returns one or more verified certificate chains given a set of roots and intermediates and would simplify this. In this use case, the returned chains can be ignored as we only care that there isn't a verification error.

Reply:
I think this is a great idea. We shouldn't be duplicating validation logic that already lives in `x509`. I've added a change to use `CertPool` and `x509.Certificate.Verify()`. It works after fixing two wrinkles:

1. the U2F root cert has an invalid `pathlen` x509 basic constraint of 0. as per RFC 5280 this means that no intermediate cert is allowed in the validation path. this isn't the first time this happens with yubico CAs, see https://labanskoller.se/blog/2019/12/30/pki-is-hard-how-yubico-trusted-openssl-and-got-it-wrong/ for a similar bug in another yubico root cert. I work around this by setting pathlen to 1 in the root cert.
2. the device attestation certs from the lucky batch do not encode X509v3 Basic Constraints. this makes `x509.Certificate.Verify()` fail. We need to set this constraint in the device cert.

Overall I'm not too surprised of these wrinkles --after all, `x509.Certificate.Verify()` performs a more strict validation conforming to RFC 5280.